### ISSUE TITLE
.NET Native: Client side evaluation of queries triggered on Release mode

### ISSUE THREAD

**Issue by peterjuras** [aspnet/EntityFramework#6381]
### Steps to reproduce
1. Clone [this sample made for reproduction](https://github.com/peterjuras/efcore-netnative-clienteval)
2. Ensure that .NET Native is disabled in project settings and run the app in debug (tested on x64)
3. Press all three buttons => No warnings and exceptions should appear
4. Stop the app and enable .NET Native in project settings (still DEBUG, still x64)
5. Run the app again and press the buttons => Exceptions are thrown for the WHERE Clause in search
### The issue

Some queries that get translated to a server side evaluated SQL statement without the .NET Native Toolchain enabled are no longer translated when the .NET Native Toolchain is enabled. The sample from above throws the following error:

Exception message:

```
Warning as error exception for warning 'RelationalEventId.QueryClientEvaluationWarning': The LINQ expression '([child].SomeString.Contains(__search_0) OrElse [model].SomeString.Contains(__search_1))' could not be translated and will be evaluated locally. To suppress this Exception use the DbContextOptionsBuilder.ConfigureWarnings API. ConfigureWarnings can be used when overriding the DbContext.OnConfiguring method or using AddDbContext on the application service provider.
```

<details>
<summary>

Stack trace: (Click to expand)</summary>

```
   at Microsoft.EntityFrameworkCore.Internal.InterceptingLogger`1.Log[TState](LogLevel logLevel, EventId eventId, TState state, Exception exception, Func`3 formatter)
   at Internal.Runtime.Augments.RuntimeAugments.CallDescrWorker(IntPtr callDescr) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 843
   at Internal.Runtime.TypeLoader.CallConverterThunk.InvokeTarget(Void* allocatedStackBuffer, CallConversionParameters& conversionParams) in f:\dd\ndp\fxcore\src\System.Private.TypeLoader\Internal\Runtime\TypeLoader\CallConverterThunk.cs:line 923
   at Internal.Runtime.Augments.RuntimeAugments.RawCalliHelper.Call[T](IntPtr pfn, Void* arg1, T& arg2)
   at Internal.Runtime.Augments.RuntimeAugments.RunFunctionWithConservativelyReportedBufferInternal[T](Int32 cbBuffer, IntPtr pfnTargetToInvoke, T& context, ConservativelyReportedRegionDesc& regionDesc) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 909
   at Internal.Runtime.Augments.RuntimeAugments.RunFunctionWithConservativelyReportedBuffer[T](Int32 cbBuffer, IntPtr pfnTargetToInvoke, T& context) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 891
   at Internal.Runtime.TypeLoader.CallConverterThunk.CallConversionThunk(IntPtr callerTransitionBlockParam, IntPtr callConversionId) in f:\dd\ndp\fxcore\src\System.Private.TypeLoader\Internal\Runtime\TypeLoader\CallConverterThunk.cs:line 479
   at Internal.Runtime.Augments.RuntimeAugments.CallDescrWorker(IntPtr callDescr) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 843
   at Internal.Runtime.TypeLoader.CallConverterThunk.InvokeTarget(Void* allocatedStackBuffer, CallConversionParameters& conversionParams) in f:\dd\ndp\fxcore\src\System.Private.TypeLoader\Internal\Runtime\TypeLoader\CallConverterThunk.cs:line 923
   at Internal.Runtime.Augments.RuntimeAugments.RawCalliHelper.Call[T](IntPtr pfn, Void* arg1, T& arg2)
   at Internal.Runtime.Augments.RuntimeAugments.RunFunctionWithConservativelyReportedBufferInternal[T](Int32 cbBuffer, IntPtr pfnTargetToInvoke, T& context, ConservativelyReportedRegionDesc& regionDesc) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 909
   at Internal.Runtime.Augments.RuntimeAugments.RunFunctionWithConservativelyReportedBuffer[T](Int32 cbBuffer, IntPtr pfnTargetToInvoke, T& context) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 891
   at Internal.Runtime.TypeLoader.CallConverterThunk.CallConversionThunk(IntPtr callerTransitionBlockParam, IntPtr callConversionId) in f:\dd\ndp\fxcore\src\System.Private.TypeLoader\Internal\Runtime\TypeLoader\CallConverterThunk.cs:line 479
   at Microsoft.EntityFrameworkCore.Infrastructure.SensitiveDataLogger`1.Microsoft.Extensions.Logging.ILogger.Log[TState](LogLevel logLevel, EventId eventId, TState state, Exception exception, Func`3 formatter)
   at Internal.Runtime.Augments.RuntimeAugments.CallDescrWorker(IntPtr callDescr) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 843
   at Internal.Runtime.TypeLoader.CallConverterThunk.InvokeTarget(Void* allocatedStackBuffer, CallConversionParameters& conversionParams) in f:\dd\ndp\fxcore\src\System.Private.TypeLoader\Internal\Runtime\TypeLoader\CallConverterThunk.cs:line 923
   at Internal.Runtime.Augments.RuntimeAugments.RawCalliHelper.Call[T](IntPtr pfn, Void* arg1, T& arg2)
   at Internal.Runtime.Augments.RuntimeAugments.RunFunctionWithConservativelyReportedBufferInternal[T](Int32 cbBuffer, IntPtr pfnTargetToInvoke, T& context, ConservativelyReportedRegionDesc& regionDesc) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 909
   at Internal.Runtime.Augments.RuntimeAugments.RunFunctionWithConservativelyReportedBuffer[T](Int32 cbBuffer, IntPtr pfnTargetToInvoke, T& context) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 891
   at Internal.Runtime.TypeLoader.CallConverterThunk.CallConversionThunk(IntPtr callerTransitionBlockParam, IntPtr callConversionId) in f:\dd\ndp\fxcore\src\System.Private.TypeLoader\Internal\Runtime\TypeLoader\CallConverterThunk.cs:line 479
   at Microsoft.EntityFrameworkCore.Storage.RelationalLoggerExtensions.LogWarning(ILogger logger, RelationalEventId eventId, Func`1 formatter)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.WarnClientEval(Object expression)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitWhereClause(WhereClause whereClause, QueryModel queryModel, Int32 index)
   at Remotion.Linq.Clauses.WhereClause.Accept(IQueryModelVisitor visitor, QueryModel queryModel, Int32 index)
   at Remotion.Linq.QueryModelVisitorBase.VisitBodyClauses(ObservableCollection`1 bodyClauses, QueryModel queryModel)
   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitSubQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.SqlTranslatingExpressionVisitor.VisitSubQuery(SubQueryExpression expression)
   at Remotion.Linq.Clauses.Expressions.SubQueryExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node) in f:\dd\ndp\fxcore\src\System.Linq.Expressions\System\Linq\Expressions\ExpressionVisitor.cs:line 40
   at Remotion.Linq.Parsing.ThrowingExpressionVisitor.Visit(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.SqlTranslatingExpressionVisitor.Visit(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitWhereClause(WhereClause whereClause, QueryModel queryModel, Int32 index)
   at Remotion.Linq.Clauses.WhereClause.Accept(IQueryModelVisitor visitor, QueryModel queryModel, Int32 index)
   at Remotion.Linq.QueryModelVisitorBase.VisitBodyClauses(ObservableCollection`1 bodyClauses, QueryModel queryModel)
   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateQueryExecutor[TResult](QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](QueryModel queryModel)
   at _$ILCT$.$ILT$ReflectionDynamicInvoke$.InvokeRetOI[T0,T1](Object thisPtr, IntPtr methodToCall, ArgSetupState& argSetupState, Boolean targetIsThisCall)
   at System.InvokeUtils.CallIHelperStaticCallWithInstantiation(Object thisPtr, IntPtr methodToCall, IntPtr dynamicInvokeHelperMethod, ArgSetupState& argSetupState, Boolean isTargetThisCall, IntPtr dynamicInvokeHelperGenericDictionary)
   at System.InvokeUtils.CallDynamicInvokeMethod(Object thisPtr, IntPtr methodToCall, Object thisPtrDynamicInvokeMethod, IntPtr dynamicInvokeHelperMethod, IntPtr dynamicInvokeHelperGenericDictionary, String defaultValueString, Object[] parameters, Boolean invokeMethodHelperIsThisCall, Boolean methodToCallIsThisCall) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\System\InvokeUtils.cs:line 464
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\System\Runtime\ExceptionServices\ExceptionDispatchInfo.cs:line 61
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass19_0`1.<CompileQuery>b__0()
   at System.Func`1.Invoke()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQuery[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at Remotion.Linq.QueryableBase`1.GetEnumerator()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source) in f:\dd\ndp\fxcore\Open\src\System.Linq\src\System\Linq\Enumerable.cs:line 3098
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source) in f:\dd\ndp\fxcore\Open\src\System.Linq\src\System\Linq\Enumerable.cs:line 1005
   at EmptyQueryNativeTest.Context.GetModels(String search) in C:\Users\pjura\Desktop\EmptyQueryNativeTest\EmptyQueryNativeTest\Context.cs:line 35
   at EmptyQueryNativeTest.MainPage.SearchWithResults(Object sender, RoutedEventArgs e) in C:\Users\pjura\Desktop\EmptyQueryNativeTest\EmptyQueryNativeTest\MainPage.xaml.cs:line 53
   at Windows.UI.Xaml.RoutedEventHandler.Invoke(Object sender, RoutedEventArgs e)
   at McgInterop.Intrinsics.HasThisCall__Proc_object__TArg0__[TArg0](Object __this, IntPtr pfn, Object arg0, TArg0 arg1)
   at McgInterop.ReverseComSharedStubs.Proc_object__TArg0__[TArg0](Object __this, __vtable_IInspectable* unsafe_sender, Void* unsafe_e, IntPtr __methodPtr) in c:\Users\pjura\Desktop\EmptyQueryNativeTest\EmptyQueryNativeTest\obj\x64\Debug\ilc\intermediate\EmptyQueryNativeTest.McgInterop\SharedStubs.g.cs:line 3985
```

</details>
### Further technical details

EF Core version: 1.0.0
Operating system: Windows 10 (Anniversary Update)
Visual Studio version: VS 2015

Other details about my project setup:
I actually had another issue in my app where queries with 0 results would take way too long to execute with .NET Native enabled, but while I created a repro sample I found that this would be the first problem that I encountered.

**Comment by divega** [comment]
Another case that forces client evaluation, in this case with LIKE:  https://github.com/aspnet/EntityFramework/issues/6397. The issue contains a full repro.

**Comment by MattWhilden** [comment]
Thanks for the great repro. I see a first chance InvalidOperationException for the second two issues. I'll need to work with some folks on our side to figure our what's the best path forward.

Is this currently blocking you or do you have a workaround for this issue? 

I've opened a tracking issue in our internal DB. If it's useful later, number is 279085

**Comment by peterjuras** [comment]
I currently have a workaround for it (by writing it in plain SQL), a fix would be nice nonetheless. However, general performance of queries in release mode are still lackluster (2ms without .NET Native vs 200-300ms (!) with). See #5570

**Comment by MattWhilden** [comment]
We believe that this issue is resolved with version 5.3.0 of Microsoft.NETCore.UniversalWindowsPlatform. You'll need to upgrade to VS 2017 to take advantage of this. Details here: https://blogs.msdn.microsoft.com/dotnet/2017/01/30/announcing-net-core-net-native-and-nuget-updates-in-vs-2017-rc/

**Comment by divega** [comment]
Closing based on @MattWhilden's comment that this should be fixed in UniversalWindowsPlatform 5.3.0. I any similar issues occur using the latest version, please create a new issue.
